cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/worker/WorkerProgramRunner.java

```diff
@@ -0,0 +1,114 @@
+import co.cask.cdap.internal.app.runtime.ProgramOptionConstants;
+import co.cask.cdap.proto.ProgramType;
+import co.cask.tephra.TransactionSystemClient;
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.Service;
+import com.google.inject.Inject;
+import org.apache.twill.api.RunId;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import org.apache.twill.internal.RunIds;
+
+/**
+ * A {@link ProgramRunner} that runs a {@link Worker}.
+ */
+public class WorkerProgramRunner implements ProgramRunner {
+
+  private final CConfiguration cConf;
+  private final MetricsCollectionService metricsCollectionService;
+  private final DatasetFramework datasetFramework;
+  private final DiscoveryServiceClient discoveryServiceClient;
+  private final TransactionSystemClient txClient;
+
+  @Inject
+  public WorkerProgramRunner(CConfiguration cConf, MetricsCollectionService metricsCollectionService,
+                             DatasetFramework datasetFramework, DiscoveryServiceClient discoveryServiceClient,
+                             TransactionSystemClient txClient) {
+    this.cConf = cConf;
+    this.metricsCollectionService = metricsCollectionService;
+    this.datasetFramework = datasetFramework;
+    this.discoveryServiceClient = discoveryServiceClient;
+    this.txClient = txClient;
+  }
+
+  @Override
+  public ProgramController run(Program program, ProgramOptions options) {
+    String workerName = options.getName();
+    Preconditions.checkNotNull(workerName, "Missing worker name.");
+
+    ApplicationSpecification appSpec = program.getApplicationSpecification();
+    Preconditions.checkNotNull(appSpec, "Missing application specification.");
+
+    int instanceId = Integer.parseInt(options.getArguments().getOption(ProgramOptionConstants.INSTANCE_ID, "0"));
+    Preconditions.checkArgument(instanceId >= 0, "Missing instance Id");
+
+    int instanceCount = appSpec.getWorkers().get(program.getName()).getInstances();
+    Preconditions.checkArgument(instanceCount > 0, "Invalid or missing instance count");
+
+    RunId runId = RunIds.generate();
+
+    ProgramType programType = program.getType();
+    Preconditions.checkNotNull(programType, "Missing processor type.");
+    Preconditions.checkArgument(programType == ProgramType.WORKER, "Only Worker process type is supported.");
+
+    WorkerSpecification spec = appSpec.getWorkers().get(program.getName());
+    Preconditions.checkArgument(spec != null, "Missing Worker specification for {}", program.getId());
+
+    BasicWorkerContext context = new BasicWorkerContext(spec, program, runId, instanceId, instanceCount,
+                                                        options.getUserArguments(), cConf,
+                                                        metricsCollectionService, datasetFramework,
+                                                        txClient, discoveryServiceClient);
+    Service worker = new WorkerDriver(program, spec, context);
```

Comment:
The type can remain `WorkerDriver` since you anyway cast it again inside the adapter to change instances.